fastar/rules.py

```diff
@@ -160,38 +164,39 @@ def concatenate_dependency_rule(outbox, *operands, dimension):
 
 dependency_rules[lax.concatenate_p] = concatenate_dependency_rule
 
-def slice_dependency_rule(outbox, operand, start_indices, limit_indices, strides):
-  outstart, outshape = outbox
-  strides = np.ones_like(outshape) if strides is None else np.array(strides)
-  zeros = np.zeros_like(outshape)
+def slice_dependency_rule(outstart, outcount, operand, start_indices, limit_indices, strides):
+  strides = np.ones_like(outcount.shape) if strides is None else np.array(strides)
+  zeros = np.zeros_like(outcount.shape)
   return ([outstart * strides + start_indices],
-          [lax.pad(np.ones(outshape, int), 0, zip(zeros, zeros, strides - 1))],
+          [lax.pad(materialize(outcount), 0, zip(zeros, zeros, strides - 1))],
           lambda inslice: lax.slice(inslice, zeros, inslice.shape, strides))
 
 dependency_rules[lax.slice_p] = slice_dependency_rule
 
-def transpose_dependency_rule(outbox, operand, permutation):
-  outstart, outshape = outbox
+def transpose_dependency_rule(outstart, outcount, operand, permutation):
   inverse_perm = np.argsort(permutation)
   return ([np.take(outstart, inverse_perm)],
-          [np.ones(np.take(outshape, inverse_perm), int)],
+          [Ones(np.take(outcount.shape, inverse_perm))
+           if is_ones(outcount) else laxref.transpose(outcount, inverse_perm)],
```

Comment:
Could use `np.transpose` instead of `lax_ref.transpose` since IIRC the two are identical.